[inductor] Allow scheduler to reorder loops when it enables fusions

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89877

This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.
Fixes #93718

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

peterbell10 committed:
Fix fusion of softmax on outer dimension

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix fusion of softmax on outer dimension"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix fusion of softmax on outer dimension"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix fusion of softmax on outer dimension"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix fusion of softmax on outer dimension"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] WIP: Allow scheduler to reorder loops when it allow more fusion"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] WIP: Allow scheduler to reorder loops when it allow more fusion"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] WIP: Allow scheduler to reorder loops when it allow more fusion"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] WIP: Allow scheduler to reorder loops when it enables fusions"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] WIP: Allow scheduler to reorder loops when it enables fusions"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] WIP: Allow scheduler to reorder loops when it enables fusions"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] WIP: Allow scheduler to reorder loops when it enables fusions"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

lezcano commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
+
+        dep1_permutation = sorted(range(ndim), key=lambda i: expr_key(strides1[i]))
+        dep2_permutation = sorted(range(ndim), key=lambda i: expr_key(strides2[i]))
+        if dep1_permutation == dep2_permutation:
```
Sanity check. Is this just an optimisation as we know that if the permutations were the same, then the regular fusion algorithm would have fused the nodes, or does it do something else too?

peterbell10 replied:
It's also possible that the dependencies match but the fusion was rejected for some other reason. So we want to explicitly exclude these cases.

lezcano commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
+        if dep1.index != dep2_reindexed:
+            return None
+
+        # Permute node2 to iteration in the same order as node1
```
```suggestion
        # Permute node2 to iterate it in the same order as node1
```

lezcano commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
+                for d1, d2 in zip(dep1_permutation, dep2_permutation)
+            },
+        )
+        if dep1.index != dep2_reindexed:
+            return None
```
Should we make this an assert?

peterbell10 replied:
`stride_vars` can give the wrong answer in some edge cases (which I documented in #90184), so this check can fail for that reason.

lezcano replied:
oof. Perhaps leave a comment?

lezcano commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
+        ] = collections.defaultdict(set)
+
+        for buf_name, node_grouping in buffer_names_grouping.items():
+            for node1, node2 in self._upper_triangle_pairs(node_grouping):
```
-> combinations(node_grouping, 2)

lezcano commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
+                        continue
+                    else:
+                        node1, node2 = node2, node1
+
+                if (node1, node2) in possible_fusions or (
+                    node2,
+                    node1,
+                ) in possible_fusions:
+                    continue
```
Move before the previous swap and simplify to just `node1, node2 in possible_fusions`.

lezcano commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
+
+                if any(
+                    not isinstance(n, (FusedSchedulerNode, SchedulerNode))
+                    for n in all_nodes(node1, node2)
```
```suggestion
                    for n in chain(node1.get_nodes(), node2.get_nodes())
```

lezcano commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
+        if dep1 == dep2:
+            return None
+
+        if not isinstance(dep1, MemoryDep) or not isinstance(dep2, MemoryDep):
+            return None
```
This is filtered below, so it could go. We could even type the arguments to this function.

lezcano reviewed (commented):
Overall, it looks good. Most of the points I left are nits + a few questions.
It would really help to have a description of the algorithm in a comment.
We should also bench this.

peterbell10 committed:
Update on "[inductor] Allow scheduler to reorder loops when it enables fusions"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
+            new_node.max_order = node.max_order
+            new_node.last_usage = node.last_usage
+            new_node.written = node.written
+            return new_node
```
Both nodes will still exist for the duration of the `fuse_nodes_once` call, after which exactly one will exist. The shallow copying should be fine though as AFAIK they aren't updated during this time and these properties really should always be identicalanyway.

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
+        def sort_exprs(exprs):
+            return sorted(exprs, key=expr_key)
+
+        if sort_exprs(dep1.size) != sort_exprs(dep2.size):
```
I believe so, yes
```python
>>> x, y = sympy.symbols("x,y")
>>> str(x + y), str(y + x)
('x + y', 'x + y')
```

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -806,42 +817,51 @@ def fuse_nodes_once(self):
-            node1 = self.name_to_fused_node[node1.get_first_name()]
-            node2 = self.name_to_fused_node[node2.get_first_name()]
+
+        recent_fusions = set()
```
This is required because `get_possible_fusions()` may return a new node not in `name_to_fused_node`, so we can't trust
```
node = self.name_to_fused_node[node.get_first_name()]
```
to map an unfused nodes back to itself

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
+            return new_node
+
+        for (node1, node2), orders in reorder_opportunities.items():
+            for order in orders:
```
I changed some the names to hopefully clarify whats going on.

`orders` here is the set of all permutations returned from `pick_dependency_reorder`, but at this stage we haven't actually reordered anything.

For each permutation we then call `node.reorder_loop`  and the dependency information for the new loop is calculated in `SchedulerNode`'s constructor.

`can_fuse` then does all the usual checks, including checking that memory dependencies match which in theory we know should at least match for the dependency that `pick_dependency_reorder` returned.

We could maybe do more to prune `orders` down before trying them for real, but I'm not sure it would be worth the added code complexity.

peterbell10 committed:
Update on "[inductor] Allow scheduler to reorder loops when it enables fusions"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

Chillee commented on test/inductor/test_perf.py:
```diff
@@ -250,6 +250,14 @@ def f(a):
         inp = (T(10, 10),)
         self.assertExpectedInline(count_numel(f, *inp), """200""")
 
+    @patch.object(config, "fusion_reorder_loops", True)
```
Also, would be nice to add a bunch more tests here :)

Chillee commented on torch/_inductor/scheduler.py:
```diff
@@ -405,6 +416,54 @@ def can_inplace(self, read_dep: dependencies.MemoryDep):
+    ) -> Optional["SchedulerNode"]:
+        """Returns a new node which represents the same loop iterated in a new order
+
+        new_order represents the permutation to apply to the index variables,
```
I don't see what new_order is?

peterbell10 replied:
Ah this is out of data, it now describes the `reorder` argument.

Chillee commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +953,149 @@ def check_all_pairs(nodes):
+        def expr_key(expr):
+            # Include str(exp) to differentiate dimensions with the same value
+            # but derived from different expressions
+            return (sizevars.size_hint(expr), str(expr))
```
There's a `sympy_str` in `utils` if you really need a string.

peterbell10 replied:
I found `Expr.sort_key()` which is exactly what I want and was 1000x faster than `str(e)` for a simple 3d indexing expression.

Chillee commented on torch/_inductor/scheduler.py:
```diff
@@ -806,42 +884,57 @@ def fuse_nodes_once(self):
+
+        # recent_fusions tracks names of nodes fused during this function call.
+        #
+        # This is required because get_possible_fusions() may return a new node
```
When can this happen?

Chillee commented on torch/_inductor/scheduler.py:
```diff
@@ -850,6 +943,11 @@ def check_all_pairs(nodes):
         for node_grouping in buffer_names_grouping.values():
             check_all_pairs(node_grouping)
 
+        if config.fusion_reorder_loops:
```
Any particular reason it's structured this way rather than modifying `can_fuse`?

peterbell10 replied:
I like the fact that the reordering logic is self-contained within this one function and `can_fuse` remains relatively easy to read and understand. It also does the reordering at the ir level, so is backend agnostic.

jansel replied:
IMO we should put this in `can_fuse()`.  I think it will simplify the code (no need to iterate over O(n^2) possible fusions twice), and also be faster since I suspect reordering will be slow since it calls FX symbolic trace.

I think you can still make it self contained, basically just change:
https://github.com/pytorch/pytorch/blob/e7f1daa355064fac64e2ac9cec5d51181549ab26/torch/_inductor/scheduler.py#L1209-L1214
to:
```
if remaining_deps & node1_names: 
    return self.can_fuse_via_reordering()
```

You then need to remember to apply the reordering when the actual fusion happens.  You could do something like:
```
self._fusion_needs_reordering[(node1,node2)] = ...
```

peterbell10 replied:
> I suspect reordering will be slow since it calls FX symbolic trace.

If compilation latency is an issue I think we can re-use the original graph since in theory only the indexing formulas should change. 

> I think you can still make it self contained, basically just change: [Scheduler.can_fuse_vertical]

You also need to add support for reordering in:
- `get_backend(device).can_fuse_vertical` for all backends
- `score_fusion_memory` which is called in `can_fuse` before `can_fuse_vertical`
- `fuse_nodes_once`
 
This is not self-contained because it leaks all over the code and will make composing future transformations or adding new fusion scoring metrics significantly harder.

peterbell10 committed:
Update on "[inductor] Allow scheduler to reorder loops when it enables fusions"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

jansel commented on torch/_inductor/scheduler.py:
```diff
@@ -405,6 +416,59 @@ def can_inplace(self, read_dep: dependencies.MemoryDep):
+        reorder is a function that takes indices in the new order and
+        translates them into the indices expected by the underlying loop. e.g.
+
+            reorder = lambda index: index[1], index[0]
```
Could this just be a dict/list?  Is a callable actually needed?

peterbell10 replied:
Reverted back to a list.

jansel commented on torch/_inductor/scheduler.py:
```diff
@@ -405,6 +416,59 @@ def can_inplace(self, read_dep: dependencies.MemoryDep):
+
+        """
+
+        if len(self.get_ranges()[0]) != ndim:
```
Should this be an error?  When does this happen?

peterbell10 replied:
This is mainly for the benefit of `FusedSchedulerNode.reorder_loop` since cuda can fuse nodes that have a common numel, but not a common shape. I could move the line there if you prefer.

jansel commented on torch/_inductor/scheduler.py:
```diff
@@ -405,6 +416,59 @@ def can_inplace(self, read_dep: dependencies.MemoryDep):
+            assert len(reduce_vars) == 0
+            body = ir.LoopBody(
+                self._body,
+                [reorder(iter_vars), reduce_vars],
```
Usage of redorder doesn't seem to match what is in the docs above.

peterbell10 replied:
Good catch, the bug didn't show up with the 2d cases I was testing.

jansel commented on torch/_inductor/scheduler.py:
```diff
@@ -806,42 +889,58 @@ def fuse_nodes_once(self):
         """
         fused_nodes = set(self.nodes)
+
+        # recent_fusions tracks names of nodes fused during this function call.
+        #
+        # This is required because config.fusion_reorder_loops, when enabled,
+        # causes get_possible_fusions() to return new nodes not in
+        # name_to_fused_node, so we can't assume
+        #   node = self.name_to_fused_node[node.get_first_name()]
+        # will map an unfused nodes back to itself
+        recent_fusions: Set[str] = set()
```
Why can't we update `self.name_to_fused_node` properly?  I don't totally get why this is needed.

peterbell10 replied:
There isn't a 1-1 mapping between name and nodes, so it's impossible to represent in `name_to_fused_nodes`. We also don't want to, since the reordered nodes only exist for the duration of this function call.

jansel commented on torch/_inductor/scheduler.py:
```diff
@@ -806,42 +889,58 @@ def fuse_nodes_once(self):
         Helper to find all legal fusion opportunities, sorted by self.score_fusion()
         """
-        possible_fusions = []
+        possible_fusions = set()
```
This might make things non-deterministic since set iteration order can change run-to-run.

peterbell10 replied:
The fusions are sorted before being returned, and the set itself is only used for `.add` and `in`.

jansel replied:
I believe they are sorted by score_fusion, which can return duplicates.

peterbell10 replied:
Fair point, we could add a tie breaker to `score_fusion` like `id(node1)`?

jansel replied:
`id()` is also non-determinstic.  (It is the internal memory address.)

jansel commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +965,154 @@ def check_all_pairs(nodes):
+        Loop. Any pair of nodes in the ignore set will not be checked, on the
+        assumption these are better to fuse without reordering.
+        """
+        possible_fusions = set()
```
Danger sets are not determinstic.

jansel commented on torch/_inductor/scheduler.py:
```diff
@@ -861,6 +965,154 @@ def check_all_pairs(nodes):
+                reorder_node1 = node2.is_reduction() and not node1.is_reduction()
+                if reorder_node1:
+                    assert isinstance(node1, (SchedulerNode, FusedSchedulerNode))
+                    new_node = node1.reorder_loop(ir.inverse_reorder(order), len(order))
```
Should we cache this on the node?  Since it requires retracing the body it might be slow.

jansel reviewed (changes requested):
Awesome progress, excited to see this!

Any perf results?

peterbell10 committed:
Update on "[inductor] Allow scheduler to reorder loops when it enables fusions"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Allow scheduler to reorder loops when it enables fusions"


This allows the iteration order of pointwise loops to be changed, if it allows
for more fusions. This allows softmax on the outer dimension to be fused into a
single kernel.

[ghstack-poisoned]

peterbell10 commented:
The huggingface benchmarks show some mixed results. Compression ratio shows is greatly improved in a number of models, but with a few relatively minor regressions. Performance improved significantly in a few cases but was mostly insignificant and in the case of `LayoutLMForSequenceClassification` shows a catastrophic regression. I will need to dig into the regressions to see if we can mitigate with some better heuristics.


![image](https://user-images.githubusercontent.com/13238737/214904657-08940128-91ff-426c-ae1a-b6e5f38ef59d.png)

jansel commented:
> The huggingface benchmarks show some mixed results. Compression ratio is greatly improved in a number of models, but with a few relatively minor regressions.

IIRC higher is better for compression ratio -- so I think those are regressions.

peterbell10 commented:
I see, oh dear.

peterbell10 commented:
Okay digging into the benchmark regressions I saw that many actually generated identical code and I couldn't reproduce when benchmarking the model on its own. I've re-run the numbers without using `--diff-branch` and the results are far more sensible.

Most benchmarks show no difference at all; a few show variations in performance but when I looked into it it seems to just be noise in the benchmark. What is consistent though is the compression_ratio improvements in `BartForConditionalGeneration`, `DistillGPT2`, `ElectraForCausalLM` and `RobertaForCausalLM`.

![Screenshot from 2023-01-27 00-33-52](https://user-images.githubusercontent.com/13238737/214981084-6a65c233-dad5-41d6-adea-dd25ff974722.png)

Chillee commented:
It might be worth checking out timm models with channels last. Iirc, one of the big places where we see this kind of pattern messing up Inductor is outer-dim reductions, which you see in channels last models.

peterbell10 commented:
I ran timm with channels-last but didn't see any change outside of random variance.
```
python benchmarks/dynamo/timm_models.py --inductor --performance --amp --channels-last
```

![image](https://user-images.githubusercontent.com/13238737/215124722-e45eb078-c73a-4197-b801-f5a46361060d.png)
